pkg/controller/controller.go

```diff
@@ -702,8 +803,52 @@ func (p *csiProvisioner) volumeHandleToId(handle string) string {
 	return handle
 }
 
-// getSecretReference returns a reference to the secret specified in the given nameKey and namespaceKey parameters, or an error if the parameters are not specified correctly.
-// if neither the name or namespace parameter are set, a nil reference and no error is returned.
+func getSecretNameAndNamespaceTemplate(secret secretParamsType, storageClassParams map[string]string) (nameTemplate, namespaceTemplate string, err error) {
+	numName := 0
+	numNamespace := 0
+
+	if t, ok := storageClassParams[secret.deprecatedSecretNameKey]; ok {
+		nameTemplate = t
+		numName += 1
+		glog.Warning(deprecationWarning(secret.deprecatedSecretNameKey, secret.secretNameKey, ""))
+	}
+	if t, ok := storageClassParams[secret.deprecatedSecretNamespaceKey]; ok {
+		namespaceTemplate = t
+		numNamespace += 1
+		glog.Warning(deprecationWarning(secret.deprecatedSecretNamespaceKey, secret.secretNamespaceKey, ""))
+	}
+	if t, ok := storageClassParams[secret.secretNameKey]; ok {
+		nameTemplate = t
+		numName += 1
+	}
+	if t, ok := storageClassParams[secret.secretNamespaceKey]; ok {
+		namespaceTemplate = t
+		numNamespace += 1
+	}
+
```

Comment:
What if you had a helper function
```golang
getKeyFromParams(actor, key string) (val string, isDeprecated bool, err error)
```
that
* Returns the val corresponding to either the deprecated or the existing key if only 1 of them exists (and set `isDeprecated` accordingly)
* Returns an error if both deprecated and existing keys exist

Then the check below becomes something like:
```golang
if secretNameIsDeprecated != secretNamespaceIsDeprecated {
   throwError()
} else if secretNameIsDeprecated {
   ...
} else { // !secretNameIsDeprecated && !secretNamespaceIsDeprecated
   ...
}
```

Reply:
when passing in a key we already know whether its deprecated or not as we define that so it doesnt make sense to me to have that as a return. Also given one deprecated theres no way to associate that with the "non-deprecated" version of that key without a struct that binds both together (or some hardcoding in this function) which serves the same purpose.

Reply:
Tried to prototype this piece of logic and yeah, it doesn't look much prettier. OK to keep everything as is!

Reply:
could you LGTM?